Add document/workspace symbol request handler

ssmifi opened the pull request:
Closes https://github.com/zowe/zowe-pli-language-support/issues/119 https://github.com/zowe/zowe-pli-language-support/issues/120

- Implemented the handler for document symbol requests. The current implementation supports **procedures**, **variables**, **constants**, **structs**, and other **label**-prefixed symbols. The design allows for easy extension via the symbol builder.
- Added corresponding tests for document symbol handling.
- Additionally, updated (and simplified) the symbolic link tests to use the new `replaceNamedIndices` function.
---
- Implemented the handler for workspace symbol requests. 
- Added test for workspace symbol requests.
- Added caching for both requests (and margins and skipped code) to avoid recomputation.
---
Preview:
![image](https://github.com/user-attachments/assets/296d14eb-3271-455d-80e8-269ba65ac6c4)

montymxb commented on packages/language/src/workspace/compilation-unit.ts:
```diff
@@ -155,8 +173,7 @@ export class CompletionUnitHandler {
       }
-      skippedCode(connection, event.document.uri, unit, document);
-      marginIndicator(connection, event.document.uri, unit, document);
+      unit.requestCaches.revalidateAll({ connection, unit });
```
I presume this should be `invalidateAll`?

montymxb reviewed (commented):
General comments & feedback. Overall the cache approach appears to be a bit overkill for what we need in this regard, but a pared down version should be fine.

Additionally the calculation seems to be off, especially for synthesized elements from the pre-processor. Such as for the preprocessor *do.pli* example if you check that one out.

montymxb commented on packages/language/src/language-server/document-symbol-request.ts:
```diff
@@ -0,0 +1,347 @@
+  compilationUnit: CompilationUnit,
+  uri: URI,
+): DocumentSymbol[] {
+  if (!compilationUnit.requestCaches.get("documentSymbols").isCached()) {
```
Just revising this a bit after review. Even if the caching logic is simplified, I would still hide the cache used by the compileUnit from outside access, and abstract it away as much as possible.

ssmifi commented on packages/language/src/language-server/margin-indicator.ts:
```diff
@@ -26,25 +25,18 @@ export const MarginIndicatorNotification =
-      indicator.m = 2;
-      indicator.n = 72;
-    }
+  const cachedMargins = compilationUnit.requestCaches.get("margins").get();
```
Margins are stored per uri in the client. Besides, the rulers in vs code, which we use for the margins, are set per language, so if different units use different margins, they are all visible in all editors (unless set to no or default margins in the settings).

ssmifi commented:
> Additionally the calculation seems to be off, especially for synthesized elements from the pre-processor. Such as for the preprocessor _do.pli_ example if you check that one out.

The calculation is based on the information in the ast. For synthesized elements, there is currently no position data, but the positions for included elements seem off. I'm going to have a closer look at that and we can discuss later.

montymxb commented on packages/language/test/lsp/workspace-symbol-request.test.ts:
```diff
@@ -13,9 +13,8 @@ import { describe, test, expect } from "vitest";
-import { workspaceSymbolRequest } from "../../src/language-server/workspace-symbol-request";
 import { TextDocuments } from "../../src/language-server/text-documents";
-import { createCompilationUnit } from "../../src/workspace/compilation-unit";
+import { CompletionUnitHandler } from "../../src/workspace/compilation-unit";
```
Just a side note. I keep seeing this one, I've already corrected the naming in the plugin PR, it'll apply once that's been added

ssmifi replied:
Saw it already in your PR. 👍 
I didn't want to interfere. :)

montymxb reviewed (approved):
Alright I think we can pull this into dev now.

The outstanding aspect is related to the ranges associated with outlines, but I believe we should follow up on that separately. @msujew thoughts on that?

Assuming we pull this in, @ssmifi can you make an issue noting the current limitation on the doc/workspace symbol ranges in the outlines?

msujew commented:
@montymxb I've discussed a few things offline with @ssmifi. We will probably refactor some parts of the PR, in order to better align to the LSP types (and the lack of cross-file support in the document symbol request).

ssmifi commented:
Refactored to process only the tokens of the requested document. @montymxb please have another look when you have a moment.

- Removed all symbols without a valid range, effectively filtering out preprocessor-generated symbols.
- Removed document symbols from the cache, since it’s cleared on every document change.
  - Workspace symbols remain cached, as the request is triggered with each query update.
  - The infrastructure for document symbol caching is still in place, so re-adding support would be straightforward if needed.
    - This could be useful when switching frequently between large documents without making changes. Please let me know what you think! (cc @msujew)
- Split the document symbol request into two files for better readability.

msujew commented on packages/language/src/language-server/document-symbol-builder.ts:
```diff
@@ -0,0 +1,326 @@
+class ProcedureSymbolBuilder implements SymbolBuilder {
+  canHandle(token: IToken): boolean {
+    return (
+      token.tokenTypeIdx == CstNodeKind.EntryStatement_Variable &&
```
Suggestion: The `tokenTypeIdx` is a lexer internal variable (i.e. it is dynamically assigned while initializing the lexer) and is completely unrelated to the `CstNodeKind` enum. If you want to search for the respective enum value in the token, try `token.payload.kind` instead.

msujew commented on packages/language/src/language-server/document-symbol-builder.ts:
```diff
@@ -0,0 +1,326 @@
+class DeclareSymbolBuilder implements SymbolBuilder {
+  canHandle(token: IToken): boolean {
+    if (
+      token.tokenTypeIdx !== CstNodeKind.DefineOrdinalStatement_Unsigned0 ||
```
Suggestion: Same issue here, see above.

msujew commented on packages/language/src/language-server/document-symbol-builder.ts:
```diff
@@ -0,0 +1,326 @@
+
+    let kind: SymbolKind = SymbolKind.Variable;
+    if (
+      // Type is lower case in the ast but upper case in the definition.
+      item.attributes.some(
+        (a) =>
+          a.kind === SyntaxKind.ComputationDataAttribute &&
+          a.type?.toString() === "constant",
+      )
```
Suggestion: We perform case-insensitive parsing, so this comparison also needs to be done case-insensitive.

msujew commented on packages/language/src/language-server/document-symbol-builder.ts:
```diff
@@ -0,0 +1,326 @@
+  }
+}
+
+function isValidToken(token: IToken): boolean {
+  // Will fail for tokens with no end offset.
+  // E.g., when generated by the preprocessor.
+  return token.endOffset !== undefined && token.endOffset > 0;
+}
```
Suggestion: We already have a `isValidToken` function in `tokens.ts` which does this check.

msujew commented on packages/language/src/language-server/document-symbol-builder.ts:
```diff
@@ -0,0 +1,326 @@
+class LabelSymbolBuilder implements SymbolBuilder {
+  canHandle(token: IToken): boolean {
+    if (
+      token.tokenTypeIdx !== CstNodeKind.Reserves_Variables1 ||
```
Suggestion: Same issue here, see above.

msujew commented on packages/language/src/language-server/document-symbol-builder.ts:
```diff
@@ -0,0 +1,326 @@
+  const selectionLSPRange = rangeToLSP(textDocument, selectionRange);
+  const range = {
+    start: selectionLSPRange.start,
+    end: getLastPositionOrDefault(children, selectionLSPRange.end),
```
Suggestion: We should make use of the `END` statements (and assign a token to them) to return more accurate results where appropriate (for procedures for example). Though I'm fine with deferring this to another PR.

ssmifi replied:
Actually, the procedure builder uses the end statement now. This part is still included for the structs.

msujew replied:
Ah, I see. Yeah, that makes sense 👍

msujew commented on packages/language/src/language-server/document-symbol-builder.ts:
```diff
@@ -0,0 +1,326 @@
+    const procedureName =
+      labelPrefixStatement.labels.map((label) => label.name).join(" ") ??
+      "<unknown>";
+    const range = getSyntaxNodeRange(labelPrefixStatement.labels[0]) ?? {
+      start: 0,
+      end: 0,
+    };
```
Suggestion: We shouldn't fall back to `0:0` as the range for the syntax node. Can we use the `token.offset` position instead?

msujew commented on packages/language/src/language-server/document-symbol-builder.ts:
```diff
@@ -0,0 +1,326 @@
+    const levelSymbols: LevelSymbol[] = [];
+    const hierarchyBuilder = new LevelHierarchyBuilder();
+
+    for (const item of token.payload?.element.items.filter(
+      (i: DeclaredItem) => i.kind === SyntaxKind.DeclaredItem,
+    )) {
+      levelSymbols.push(
+        ...this.retrieveLevelSymbols(item, null, childSymbols, textDocument),
+      );
+    }
```
Suggestion: We should try to keep the amount of `any` to a minimum. Try to cast the `token.payload.element` object first, so we get better IDE support.

msujew commented on packages/language/src/language-server/document-symbol-builder.ts:
```diff
@@ -0,0 +1,326 @@
+    const documentSymbols: DocumentSymbol[] = [];
+    for (const label of labelPrefixStatement.labels) {
+      const name = label.name ?? "<unknown>";
+      const range = getSyntaxNodeRange(label) ?? { start: 0, end: 0 };
```
Suggestion: Use `token.offset` here as well.

msujew commented on packages/language/src/workspace/compilation-unit.ts:
```diff
@@ -135,6 +150,34 @@ export class CompletionUnitHandler {
+    return Array.from(this.compilationUnits.values());
+  }
+
+  getDocumentSymbols(uri: URI): DocumentSymbol[] {
+    const unit = this.getCompilationUnit(uri);
+    if (!unit) {
+      return [];
+    }
+    return documentSymbolRequest(unit, uri);
+  }
+
+  getWorkspaceSymbols(query: string): SymbolInformation[] {
+    const units = this.getAllCompilationUnits();
+    return units.flatMap((unit) => {
+      if (!unit.requestCaches.get("workspaceSymbols")) {
+        unit.requestCaches.set(
+          "workspaceSymbols",
+          workspaceSymbolRequestForCompilationUnit(this, unit),
+        );
+      }
+      const symbols = unit.requestCaches.get("workspaceSymbols") ?? [];
+      return symbols.filter((symbol) =>
+        symbol.name.toLowerCase().includes(query.toLowerCase()),
+      );
+    });
+  }
```
Suggestion: The `CompletionUnitHandler` was designed as a storage for `CompilationUnit` in addition to containing the lifecycle/update logic. I wouldn't try to also add LSP logic in there. Can we refactor this out of the handler?

ssmifi replied:
The parts regarding caching were refactored into the CU after the last review suggestions. For me, both is fine. We should just align w.r.t. this question.

ssmifi replied:
As discussed, moved to the LS.

msujew reviewed (commented):
Works pretty well. However, there are a few issues wrt. maintainability in the code.

msujew commented on packages/language/src/language-server/document-symbol-builder.ts:
```diff
@@ -0,0 +1,329 @@
+    const documentSymbols: DocumentSymbol[] = [];
+    for (const label of labelPrefixStatement.labels) {
+      const name = label.name ?? "<unknown>";
+      const range = getSyntaxNodeRange(label) ?? {
+        start: token.startOffset,
+        end: token.endOffset ?? token.startOffset,
+      };
+
+      documentSymbols.push(
+        createDocumentSymbol(
+          name,
+          SymbolKind.Key,
+          range,
+          childSymbols,
+          textDocument,
+        ),
+      );
```
Suggestion: Thinking about this again, will `getSyntaxNodeRange` ever return `undefined` for a valid token? And if the token is invalid, shouldn't we prevent the creation of the document symbol? IMO:
1. `getSyntaxNodeRange` should also check for `isValidToken` and return `undefined` if the token is not valid
2. If `getSyntaxNodeRange` returns `undefined`, skip the respective element, since we cannot navigate to it anyway.

msujew commented on packages/language/src/language-server/document-symbol-builder.ts:
```diff
@@ -0,0 +1,329 @@
+class ProcedureSymbolBuilder implements SymbolBuilder {
+  canHandle(token: IToken): boolean {
+    return (
+      token.payload.kind == CstNodeKind.ProcedureStatement_PROCEDURE &&
+      token.payload?.element.kind === SyntaxKind.ProcedureStatement
```
Minor: Either test for `undefined` on none of the payloads, or only the first one. The second test cannot be undefined, if the first is already defined.

msujew reviewed (commented):
Looks good, just one more thing wrt. invalid token ranges. See below.

msujew reviewed (approved):
Thank you, works great!

ssmifi commented:
Thanks again for the thorough review!
